phpBB/phpbb/style/manager.php

```diff
@@ -0,0 +1,436 @@
+		{
+			if (in_array($dir, $installed_dirs))
+			{
+				// Style is already installed
+				continue;
+			}
+			$cfg = $this->read_style_cfg($dir);
+			if ($cfg === false)
+			{
+				// Invalid style.cfg
+				continue;
+			}
+
+			// Style should be available for installation
+			$parent = $cfg['parent'];
+			$style = array(
+				'style_id'			=> 0,
+				'style_name'		=> $cfg['name'],
+				'style_copyright'	=> $cfg['copyright'],
+				'style_active'		=> 0,
+				'style_path'		=> $dir,
+				'bbcode_bitfield'	=> $cfg['template_bitfield'],
+				'style_parent_id'	=> 0,
+				'style_parent_tree'	=> '',
+				// Extra values for styles list
+				// All extra variable start with _ so they won't be confused with data that can be added to styles table
+				'_inherit_name'			=> $parent,
+				'_available'			=> true,
+				'_note'					=> '',
+			);
+
+			// Check style inheritance
+			if ($parent != '')
+			{
+				if (isset($installed_names[$parent]))
+				{
+					// Parent style is installed
+					$row = $installed_names[$parent];
+					$style['style_parent_id'] = $row['id'];
+					$style['style_parent_tree'] = $row['tree'];
+				}
+				else
+				{
+					// Parent style is not installed yet
+					$style['_available'] = false;
+					$style['_note'] = sprintf($this->user->lang['REQUIRES_STYLE'], htmlspecialchars($parent));
+				}
+			}
+
+			if ($all || $style['_available'])
+			{
+				$styles[] = $style;
+			}
+		}
+
+		return $styles;
+	}
+
+	/**
+	* Find all directories that have styles
+	*
+	* @return array Directory names
+	*/
+	protected function find_style_dirs()
+	{
+		$styles = array();
+
+		$dp = @opendir($this->styles_path);
+		if ($dp)
+		{
+			while (($file = readdir($dp)) !== false)
+			{
+				$dir = $this->styles_path . $file;
+				if ($file[0] == '.' || !is_dir($dir))
+				{
+					continue;
+				}
+
+				if (file_exists("{$dir}/style.cfg"))
+				{
+					$styles[] = $file;
+				}
+			}
+			closedir($dp);
+		}
+
+		return $styles;
+	}
+
+	/**
+	* Lists all styles
+	*
+	* @return array Rows with styles data
+	*/
+	public function get_installed_styles()
+	{
+		$sql = 'SELECT *
+			FROM ' . $this->styles_table;
+		$result = $this->db->sql_query($sql);
+
+		$rows = $this->db->sql_fetchrowset($result);
+		$this->db->sql_freeresult($result);
+
+		return $rows;
+	}
+
+	public function get_style_data($field, $value)
+	{
+		// TODO: Review this, possible security issue
+		// if not, maybe field doesnt need to be escaped
+		$sql = "SELECT *
+			FROM " . $this->styles_table . "
+			WHERE " . $this->db->sql_escape($field) . " = '" . $this->db->sql_escape($value) . "'";
+		$result = $this->db->sql_query($sql);
+
+		$data = $this->db->sql_fetchrow($result);
+		$this->db->sql_freeresult($result);
+
+		return $data;
+	}
+
+	/**
+	* Read style configuration file
+	*
+	* @param string $dir style directory
+	* @return array|bool Style data, false on error
+	*/
+	public function read_style_cfg($dir)
+	{
+		static $required = array('name', 'phpbb_version', 'copyright');
```

Comment:
to be stored in a class constant (or static as constant array are not allowed in our supported php version)